File: day1/solution1.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>
#include <cmath>
#include <string>
#include <utility>

class Solution {
private:
    std::string input_file;

public:
    Solution(const std::string& file = "input1.txt") : input_file(file) {}

    std::pair<std::vector<int>, std::vector<int> > read_input() {
        std::ifstream file(input_file);
        std::vector<int> list1, list2;

        if (file.is_open()) {
            int a, b;
            while (file >> a >> b) {
                list1.push_back(a);
                list2.push_back(b);
            }
            file.close();
        }
        return std::make_pair(list1, list2);
    }

    int get_distance() {
        std::pair<std::vector<int>, std::vector<int> > data = read_input();
        std::vector<int>& list1 = data.first;
        std::vector<int>& list2 = data.second;

        std::sort(list1.begin(), list1.end());
        std::sort(list2.begin(), list2.end());

        int distance = 0;
        for (size_t i = 0; i < list1.size(); ++i) {
            distance += std::abs(list1[i] - list2[i]);
        }

        return distance;
    }
};

int main() {
    Solution solution("input1.txt");
    std::cout << "Total Distance: " << solution.get_distance() << std::endl;
    return 0;
}

File: day1/solution1.py
from collections import Counter

class Solution:
    def read_input(self):
        with open("input1.txt", "r") as file:
            list1, list2 = zip(*(map(int, line.split()) for line in file))
        return list(list1), list(list2)

    def get_distance(self):
        data1, data2 = self.read_input()
        data1.sort()
        data2.sort()

        return sum(abs(a - b) for a, b in zip(data1, data2))

    def get_similarity(self):
        data1, data2 = self.read_input()
        data2_counts = Counter(data2)
        print(data2_counts)

        similarity = 0
        for i in range(len(data1)):
            similarity += data1[i] * data2_counts[data1[i]]
        return similarity


print(Solution().get_distance())
print(Solution().get_similarity())